src/events/provider.py

```diff
@@ -44,38 +44,45 @@ def on_subject_requested(self, event: SubjectRequestedEvent):
         if not self.charm.unit.is_leader():
             return
 
-        extra_user_roles = event.extra_user_roles or ""
-        subject = event.subject or ""
         relation = event.relation
         username = f"relation-{relation.id}"
-        password = (
-            self.charm.context.cluster.client_passwords.get(username)
-            or self.charm.workload.generate_password()
-        )
+        password = self.charm.context.cluster.client_passwords.get(username, "")
+
+        # All units can update their own authfile. If password is not yet set, wait until
+        # leader creates the password.
+        if not password and self.charm.unit.is_leader():
+            password = self.charm.workload.generate_password()
+        if not password and not self.charm.unit.is_leader():
+            event.defer()
+            return
+
+        extra_user_roles = event.extra_user_roles or ""
+        subject = event.subject or ""
         endpoints = self.charm.context.endpoints
         tls = "enabled" if self.charm.context.cluster.tls_enabled else "disabled"
 
         self.charm.auth_manager.add_user(username=username, password=password)
         self.charm.auth_manager.add_acl(username=username, subject=subject, role=extra_user_roles)
         self.charm.auth_manager.write_authfile()
 
-        # non-leader units need cluster_config_changed event to update their super.users
-        self.charm.context.cluster.update(
-            {username: password, "super-users": self.charm.context.super_users}
-        )
+        # non-leader units need cluster_config_changed event to update their authfiles
+        if self.charm.unit.is_leader():
```

Comment:
**niptick** similarly as before, if we are already a leader (based on the first if condition), this seems to not be needed. Also in general, I would nudge to a pattern where we fist state condition to return or defer, and then carry that assumption through, with no need of if statements and indentation in the main happy flow.

Reply:
L44 needs to be removed. Will update.

Reply:
Yes, that's fine! I got a bit confused by the code, and I wasn't sure what should have been the correct logic